simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,427 @@
+
+        # Either use the supplied irls_threshold, or fix base on distribution of
+        # model values
+        for reg, norms in zip(self.reg.objfcts, self.metrics.input_norms):
+            if not isinstance(reg, Sparse):
+                continue
+
+            for obj in reg.objfcts:
+                threshold = np.percentile(
+                    np.abs(obj.mapping * obj._delta_m(self.invProb.model)),
+                    self.percentile,
+                )
+                if isinstance(obj, SmoothnessFirstOrder):
+                    threshold /= reg.regularization_mesh.base_length
+
+                obj.irls_threshold = threshold
+
+            reg.norms = norms
+
+            if self.verbose:
+                print("irls_threshold " + str(reg.objfcts[0].irls_threshold))
+
+        # Save l2-model
+        self.invProb.l2model = self.invProb.model.copy()
+
+    def validate(self, directiveList=None):
+        directive_list = directiveList.dList
+        self_ind = directive_list.index(self)
+        lin_precond_ind = [isinstance(d, UpdatePreconditioner) for d in directive_list]
+
+        if any(lin_precond_ind):
+            assert lin_precond_ind.index(True) > self_ind, (
+                "The directive 'UpdatePreconditioner' must be after Update_IRLS "
+                "in the directiveList"
+            )
+        else:
+            warnings.warn(
+                "Without a Linear preconditioner, convergence may be slow. "
+                "Consider adding `Directives.UpdatePreconditioner` to your "
+                "directives list",
+                stacklevel=2,
+            )
+
+        beta_schedule = [
+            d for d in directive_list if isinstance(d, BetaSchedule) and d != self
+        ]
+
+        assert len(beta_schedule) == 0, (
+            "Beta scheduling is handled by the `UpdateIRLS` directive."
+            "Remove the redundant `BetaSchedule` from your list of directives.",
+        )
+
+        spherical_scale = [isinstance(d, SphericalDomain) for d in directive_list]
+        if any(spherical_scale):
+            assert spherical_scale.index(True) < self_ind, (
+                "The directive 'SphericalDomain' must be before UpdateIRLS "
+                "in the directiveList"
+            )
+
+        return True
+
+    def stopping_criteria(self):
+        """
+        Check for stopping criteria of max_irls_iteration or minimum change.
+        """
+        phim_new = 0
+        for reg in self.reg.objfcts:
+            if isinstance(reg, (Sparse, BaseSparse)):
+                reg.model = self.invProb.model
+                phim_new += reg(reg.model)
+
+        # Check for maximum number of IRLS cycles
+        if self.metrics.irls_iteration_count == self.max_irls_iterations:
+            if self.verbose:
+                print(
+                    "Reach maximum number of IRLS cycles:"
+                    + " {0:d}".format(self.max_irls_iterations)
```

Comment:
Small change here to use f-string.
```suggestion
                    "Reach maximum number of IRLS cycles:"
                    + f" {self.max_irls_iterations:d}"
```